### DISCUSSION TITLE
Could you provide a comprehensive tutorial on compiling in Linux?

### DISCUSSION THREAD

**Discussion by brewswang** [baresip/baresip#3215] [Q&A]
能给一个完整的Linux编译教程吗？谢谢。
Could you provide a comprehensive tutorial on compiling in Linux? Thanks!

**Comment by sreimers** [comment]
https://github.com/baresip/baresip/wiki/Install:-Stable-Release

**Comment by brewswang** [comment]
Have you tried it? Can it really compile successfully by following the documentation you provided?

**Comment by brewswang** [comment]
-- Checking for one of the modules 'fdk-aac'
-- Could NOT find AAC (missing: AAC_LIBRARY AAC_INCLUDE_DIR) 
-- Could NOT find AMR (missing: AMR_FOUND) 
-- Checking for one of the modules 'aptx'
-- Could NOT find APTX (missing: APTX_LIBRARY APTX_INCLUDE_DIR) 
-- Could NOT find AOM (missing: AOM_LIBRARY AOM_INCLUDE_DIR) 
-- Could NOT find FFMPEG (missing: FFMPEG_AVCODEC_LIBRARIES FFMPEG_AVCODEC_INCLUDE_DIRS avcodec avfilter avformat swscale avdevice avutil) 
-- Could NOT find FFMPEG (missing: FFMPEG_AVCODEC_LIBRARIES FFMPEG_AVCODEC_INCLUDE_DIRS avcodec avfilter avformat swscale swresample avdevice avutil) 
-- Could NOT find FFMPEG (missing: FFMPEG_AVCODEC_LIBRARIES FFMPEG_AVCODEC_INCLUDE_DIRS avcodec avfilter avformat swscale swresample avdevice avutil) 
-- Checking for one of the modules 'codec2'
-- Could NOT find CODEC2 (missing: CODEC2_LIBRARY CODEC2_INCLUDE_DIR) 
-- Could NOT find DIRECTFB (missing: DIRECTFB_LIBRARIES DIRECTFB_INCLUDE_DIR) 
-- Could NOT find SPANDSP (missing: SPANDSP_LIBRARY SPANDSP_INCLUDE_DIR) 
-- Checking for one of the modules 'g7221'
-- Could NOT find G7221 (missing: G7221_LIBRARY G7221_INCLUDE_DIR) 
-- Could NOT find SPANDSP (missing: SPANDSP_LIBRARY SPANDSP_INCLUDE_DIR) 
-- Could NOT find GST (missing: GST_FOUND) 
-- Checking for module 'gtk+-3.0'
--   No package 'gtk+-3.0' found
-- Could NOT find GTK3 (missing: GTK3_LIBRARIES GTK3_CFLAGS GTK3_CFLAGS_OTHER) 
-- Could NOT find GZRTP (missing: GZRTP_LIBRARY GZRTP_INCLUDE_DIR) 
-- Checking for one of the modules 'jack'
-- Could NOT find JACK (missing: JACK_LIBRARY JACK_INCLUDE_DIR) 
-- Checking for one of the modules 'twolame'
-- Checking for one of the modules 'mpg123'
-- Checking for one of the modules 'speexdsp'
-- Could NOT find MPA (missing: MPA_LIBRARIES MPA_INCLUDE_DIRS) 
-- Could NOT find MOSQUITTO (missing: MOSQUITTO_LIBRARY MOSQUITTO_INCLUDE_DIR) 
-- libmosquitto include dir: MOSQUITTO_INCLUDE_DIR-NOTFOUND
-- libmosquitto: MOSQUITTO_LIBRARY-NOTFOUND
-- Could NOT find OPENSLES (missing: OPENSLES_LIBRARY OPENSLES_INCLUDE_DIR) 
-- Checking for one of the modules 'opus'
-- Could NOT find OPUS (missing: OPUS_LIBRARY) 
-- Checking for one of the modules 'opus'
-- Could NOT find OPUS (missing: OPUS_LIBRARY) 
-- Checking for one of the modules 'libpipewire-0.3'
-- Could NOT find PIPEWIRE (missing: PIPEWIRE_INCLUDE_DIRS PIPEWIRE_LIBRARIES) 
-- Could NOT find SPANDSP (missing: SPANDSP_LIBRARY SPANDSP_INCLUDE_DIR) 
-- Checking for one of the modules 'portaudio'
-- Could NOT find PORTAUDIO (missing: PORTAUDIO_LIBRARY PORTAUDIO_INCLUDE_DIR) 
-- Checking for one of the modules 'pulse'
-- Checking for one of the modules 'sdl2'
-- Could NOT find SDL (missing: SDL_LIBRARY SDL_INCLUDE_DIR) 
-- Could NOT find PNG (missing: PNG_LIBRARY PNG_INCLUDE_DIR) 
-- Checking for module 'sndfile'
--   No package 'sndfile' found
-- Could NOT find SNDFILE (missing: SNDFILE_INCLUDE_DIR SNDFILE_LIBRARIES) 
-- Could NOT find FFMPEG (missing: FFMPEG_AVCODEC_LIBRARIES FFMPEG_AVCODEC_INCLUDE_DIRS avcodec avfilter avformat swscale swresample avdevice avutil) 
-- Checking for module 'libv4l2'
--   No package 'libv4l2' found
-- Could NOT find V4L2 (missing: V4L2_LIBRARY V4L2_INCLUDE_DIR) 
-- Could NOT find VPX (missing: VPX_LIBRARY VPX_INCLUDE_DIR) 
-- Could NOT find VPX (missing: VPX_LIBRARY VPX_INCLUDE_DIR) 
-- Could NOT find WEBRTC_AEC (missing: WEBRTC_AEC_LIBRARY WEBRTC_AEC_INCLUDE_DIR) 
-- Could NOT find WEBRTC_AECM (missing: WEBRTC_AECM_LIBRARY WEBRTC_AECM_INCLUDE_DIR)

**Comment by GitHubIsShortOnUsernames** [comment]
You need to install the needed dependencies. On Debian the required packages are usually named libXYZ-dev, e.g. libspandsp-dev or libopus-dev. On Redhat based distros it's XYZ-devel, e.g. opus-devel. The important part is that you always need the "dev" or "devel" package to also install the required includes. The runtime packages are automatically pulled as dependencies of the development ones.

**Comment by robert-scheck** [comment]
Yes, the documentation does not cover every aspect of every random Linux distribution out there. But, if you would like to compile software on your own, knowledge about your own Linux distribution is appreciated. Some people here might be able to provide help for a specific Linux distribution, but for that it would be helpful, if you provide information about the Linux distribution you are using and which features/modules of baresip you would like to use. Just because various stuff can not be found (as shown in the messages in your provided output), this doesn't necessarily mean baresip doesn't compile or work. Baresip is modular, has many modules and only very few people really need *all* modules. For example, if you don't care about JACK, you can ignore absence of JACK libraries/headers.

**Comment by timmrogers** [comment]
I do think that adding information on the dependencies for various Linux
distributions to the documentation would be helpful.

Here are my notes on compiling with Debian 12, in case that is useful to
anyone. I still see some warnings when I compile, but this meets my needs,
and works on both AMD64 and ARM64 devices.

###

Compiling it is a bit of a process, and requires a substantial number of
libraries. From a clean Debian Linux 12 installation, install all of the
following:

apt install build-essential
apt install cmake
apt install clang

apt install zlib1g-dev
apt install libssl-dev
apt install libspandsp*
apt install libspandsp-dev
apt install python3-pkgconfig

apt install libpipewire*
apt install libcodec*
apt install libaacs-dev
apt install libfaad*
apt install libopencore-amr*
apt install libaom*
apt install libmosquitto*
apt install libsndfile*
apt install libjack-jackd2-dev
apt install libpng-dev
apt install libpulse-dev
apt install portaudio19-dev
apt install libavcodec-dev
apt install libavfilter-dev
apt install libavdevice-dev
apt install libopenaptx*

apt install libfdk-aac-dev
apt install libv4l-dev
apt install libdirectfb-dev
apt install libvpx-dev

apt install libwebrtc-audio-processing-dev
apt install libsdl2-dev
apt install libgtk-3-dev
apt install libgtk-4-dev
apt install libgstreamer1.0-dev libgstreamer-plugins-base1.0-dev
libgstreamerd-3-dev
apt install libccrtp-dev

(These may not be necessary)

apt install libspeexdsp1
apt install libspeexdsp-dev libspeex-dev
apt install libopenh264-dev
apt install libbzrtp-dev libbzrtp0
